hazelcast/src/test/java/com/hazelcast/cluster/TcpIpMemberConnectionLossTest.java

```diff
@@ -0,0 +1,155 @@
+ * limitations under the License.
+ */
+
+package com.hazelcast.cluster;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.config.JoinConfig;
+import com.hazelcast.config.NetworkConfig;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.instance.impl.HazelcastInstanceFactory;
+import com.hazelcast.test.Accessors;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.annotation.SlowTest;
+import org.jetbrains.annotations.NotNull;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+
+import static com.hazelcast.spi.properties.ClusterProperty.CONNECTION_MONITOR_INTERVAL;
+import static com.hazelcast.test.HazelcastTestSupport.assertClusterSize;
+import static com.hazelcast.test.HazelcastTestSupport.assertTrueEventually;
+import static com.hazelcast.test.HazelcastTestSupport.randomName;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(SlowTest.class)
+public class TcpIpMemberConnectionLossTest {
+
+    @Before
+    @After
+    public void killAllHazelcastInstances() throws IOException {
+        HazelcastInstanceFactory.terminateAll();
+    }
+
+    @Test
+    public void test_whenCanNotReconnect() {
+        Config config = getConfig();
+
+        HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
+        HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
+        assertClusterSize(2, h1, h2);
+
+        //when a member disappears and the other one can't connect to it
+        h1.getLifecycleService().terminate();
+        //then it eventually gets dropped from the cluster
+        assertTrueEventually(() -> assertClusterSize(1, h2));
+    }
+
+    @Test
+    public void test_whenConnectionKeepsDying() throws Exception {
+        Config config = getConfig();
+
+        HazelcastInstance h1 = Hazelcast.newHazelcastInstance(config);
+        HazelcastInstance h2 = Hazelcast.newHazelcastInstance(config);
+        assertClusterSize(2, h1, h2);
+
+        InetSocketAddress address = Accessors.getNodeEngineImpl(h1).getThisAddress().getInetSocketAddress();
+        RejectionProxy rejectionProxy = new RejectionProxy(address.getHostName(), address.getPort());
+        try {
+            //when a member disappears
+            h1.getLifecycleService().terminate();
+            //and connections to it don't succeed (they can be established, but immediately drop)
+            //(for example a TCP proxy might do this)
+            rejectionProxy.start();
+
+            //then it eventually gets dropped from the cluster
+            assertTrueEventually(() -> assertClusterSize(1, h2));
```

Comment:
Kudos for figuring out a way to test this issue with the `RejectionProxy`. One issue is that the test doesn't fail with current `master` behaviour, because after 60 seconds the heartbeat timeout kicks in and removes the gone member -> assertion passes. So in order to make sure this fails in case of regression, I would prolong heartbeat timeout and add an explicit, lower timeout to `assertTrueEventually` eg:
```
config.setProperty(MAX_NO_HEARTBEAT_SECONDS.getName(), "120");
...
assertTrueEventually(() -> assertClusterSize(1, h2), 45);
```

Reply:
Ok, will do.